core/src/cpu/air/mod.rs

```diff
@@ -506,8 +541,53 @@ impl CpuChip {
         // If we're halting and it's a transition, then the next.is_real should be 0.
         builder
             .when_transition()
-            .when(is_halt + local.selectors.is_unimpl)
+            .when(is_halt.clone() + local.selectors.is_unimpl)
             .assert_zero(next.is_real);
+
+        builder.when(is_halt.clone()).assert_zero(local.next_pc);
+    }
+
+    /// Constraint related to the public values.
+    pub(crate) fn public_values_eval<AB: SP1AirBuilder>(
+        &self,
+        builder: &mut AB,
+        local: &CpuCols<AB::Var>,
+        next: &CpuCols<AB::Var>,
+        public_values: &PublicValues<Word<AB::Expr>, AB::Expr>,
+        is_halt: AB::Expr,
+    ) {
+        // Verify the public value's shard.
+        builder
+            .when(local.is_real)
+            .assert_eq(public_values.shard.clone(), local.shard);
+
+        // Verify the public value's first row pc.
+        builder
+            .when_first_row()
+            .assert_eq(public_values.first_row_pc.clone(), local.pc);
+
+        // If the last real row is not the last row, verify the public value's next pc.
+        builder
+            .when_transition()
+            .when(local.is_real - next.is_real)
+            .assert_eq(public_values.last_row_next_pc.clone(), local.next_pc);
```

Comment:
comment is a bit confusing to me. suggestion: "If this row is the last real row, verify public_values.last_row_next_pc".

Also, do we not have to worry about when local.is_real = 0 and next.is_real = 1? seems like that shouldn't happen because first row probably needs to be real?

Reply:
We don't have to worry about that.  There are constraints that check that there is at most 1 at_real transition and that at_real starts at 1.